src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java

```diff
@@ -0,0 +1,279 @@
+     * {@link ProductType}/{@link ProductTypeDraft} instances.
+     *
+     * @param productTypes list of {@link ProductType}/{@link ProductTypeDraft}
+     * @param <T>          a type that extends of {@link WithKey}.
+     * @return the map of keys to {@link ProductType}/{@link ProductTypeDraft} instances.
+     */
+    protected <T extends WithKey> Map<String, T> getKeysProductTypeMap(@Nonnull final List<T> productTypes) {
+        return productTypes.stream().collect(Collectors.toMap(WithKey::getKey, p -> p,
+            (productTypeA, productTypeB) -> productTypeB));
+    }
+
+    /**
+     * Given a {@link String} {@code errorMessage} and a {@link Throwable} {@code exception}, this method calls the
+     * optional error callback specified in the {@code syncOptions} and updates the {@code statistics} instance by
+     * incrementing the total number of failed product types to sync.
+     *
+     * @param errorMessage The error message describing the reason(s) of failure.
+     * @param exception    The exception that called caused the failure, if any.
+     * @param failedTimes  The number of times that the failed product types counter is incremented.
+     */
+    private void handleError(@Nonnull final String errorMessage, @Nullable final Throwable exception,
+                             final int failedTimes) {
+
+        syncOptions.applyErrorCallback(errorMessage, exception);
+        statistics.incrementFailed(failedTimes);
+    }
+
+    /**
+     * Given a list of product type drafts, attempts to sync the drafts with the existing products types in the CTP
+     * project. The product type and the draft are considered to match if they have the same key.
+     *
+     * @param oldProductTypes old product types.
+     * @param newProductTypes drafts that need to be synced.
+     * @return a future which contains an empty result after execution of the update
+     */
+    private CompletionStage<ProductTypeSyncStatistics> syncBatch(
+            @Nonnull final List<ProductType> oldProductTypes,
+            @Nonnull final List<ProductTypeDraft> newProductTypes) {
+
+        final Map<String, ProductType> oldProductTypeMap = getKeysProductTypeMap(oldProductTypes);
+
+        List<CompletableFuture<Void>> futures = newProductTypes
+                .stream()
+                .map(newProductType -> {
+                    final ProductType oldProductType = oldProductTypeMap.get(newProductType.getKey());
+
+                    if (oldProductType == null) {
+                        return createProductType(newProductType);
+                    } else {
+                        return updateProductType(oldProductType, newProductType);
+                    }
+                })
+                .map(CompletionStage::toCompletableFuture)
+                .collect(toList());
+
+        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]))
+                .thenApply(result -> statistics);
```

Comment:
what do you think if we can refactor this to return in one step: 

````java
    private CompletionStage<ProductTypeSyncStatistics> syncBatch(
            @Nonnull final List<ProductType> oldProductTypes,
            @Nonnull final List<ProductTypeDraft> newProductTypes) {

        final Map<String, ProductType> oldProductTypeMap = getKeysProductTypeMap(oldProductTypes);

        return CompletableFuture.allOf(newProductTypes
            .stream()
            .map(newProductType -> {
                final ProductType oldProductType = oldProductTypeMap.get(newProductType.getKey());

                return ofNullable(oldProductType)
                    .map(productType -> updateProductType(oldProductType, newProductType))
                    .orElseGet(() -> createProductType(newProductType));
            })
            .map(CompletionStage::toCompletableFuture)
            .toArray(CompletableFuture[]::new)).thenApply(result -> statistics);
    }
````

Reply:
a3dc4dd7cd01e75ed24e2ae0a02cd125c084b967